### PR TITLE
Add DispatchingEntityLookup

### PR DESCRIPTION
**Pull request by manicki**
DispatchingEntityLookup picks the right EntityLookup object to perform the lookup depending on the repository the input EntityId object is coming from.

Issues to be solved:
- how do we want to name such services (this is the first of the whole bunch): should they be "Multiplexing" services, or rather "Dispatching" services, or maybe something else? Any suggestions? - edit 12.10.2016: per Jeroen's suggestion "Dispatching" is used in the service name, not the original "Multiplexing"
- ~~depends on DM 6.2 which has not been tagged yet,~~
- ~~I am not quite sure how it should deal with the "main" repository, ie. how should it handle "local" EntityId (ie. not coming from the foreign repository). Should the constructor get a single "default" EntityLookup for all non-foreign EntityIds, and then a (possibly empty) list of repository speciific lookups? Or maybe rather should it expect to only get an array of lookups with special empty-string key reserved for the default lookup? The latter seems far more hacky, so I guess it is either the former approach or something even better I couldn't think of. Suggestions?~~
- to make review easier amendments to the original commit have been submitted as separate commits. They all need to be squashed into single commit before merging.

~~Note: this includes https://github.com/wmde/WikibaseDataModelServices/pull/150/commits/336a8341e7db350ec72b31696ea845167260daa8 from https://github.com/wmde/WikibaseDataModelServices/pull/154~~

Ticket: [T146989](https://phabricator.wikimedia.org/T146989)

### CODE REVIEW

File: src/Lookup/MultiplexingEntityLookup.php
Comment (jakobw) [inline src/Lookup/MultiplexingEntityLookup.php]: I think that's supposed to say "as keys"?

File: src/Lookup/MultiplexingEntityLookup.php
Comment (jakobw) [inline src/Lookup/MultiplexingEntityLookup.php]: > bee

:bee:?

File: src/Lookup/MultiplexingEntityLookup.php
Comment (jakobw) [inline src/Lookup/MultiplexingEntityLookup.php]: Might be a matter of personal preference but I think it would read better with a private method that takes care of picking the correct repository which can be used here and in `hasEntity` below.

**Review by jakobw — COMMENTED** [review]
Looks like a solid approach to me!

I like the idea of having the local lookup injected as a constructor argument and then asking the EntityId whether it's local or not before using the repo name for multiplexing.

**Comment by manicki** [comment]
Thanks! I've changed it to have local and foreign lookups separated. It makes sense for me (at least now :). Also did other things you suggested.
For the record: Based from the experience from my previous PRs I've submitted changes as separate commits. They will be squashed into single one when the merge time comes.

**Review by jakobw — APPROVED** [review]
I think the public methods probably still need `@since` tags, other than that it looks good to me!

**Comment by manicki** [comment]
Actually, the new class should have @since (and @license) docs. Somehow I managed to not include them here. Fixed now, thanks!

File: src/Lookup/MultiplexingEntityLookup.php
Comment (jakobw) [inline src/Lookup/MultiplexingEntityLookup.php]: This method and the ones using it need the `@throws` tag

**Review by jakobw — COMMENTED** [review]
One small thing missing but the rest looks good to me!

File: src/Lookup/MultiplexingEntityLookup.php
Comment (JeroenDeDauw) [inline src/Lookup/MultiplexingEntityLookup.php]: It's not so nice to force clients to deal with both as they are caused by very similar scenarios. Do they have a common supertype that can be used here instead?

File: src/Lookup/MultiplexingEntityLookup.php
Comment (JeroenDeDauw) [inline src/Lookup/MultiplexingEntityLookup.php]: Since I was fuzzy on what exactly multiplexing means, I had to look it up (not good). This service is not combining any of its input, so I think the name is misleading.

We have similar services where we used something like `DispatchingFooService` or `DelegatingFooService`. Would be nice to include a hint that this is done based on repository name.

File: tests/unit/Lookup/MultiplexingEntityLookupTest.php
Comment (JeroenDeDauw) [inline tests/unit/Lookup/MultiplexingEntityLookupTest.php]: Happy to see you using `$this->any()`. What about the method part though - is that needed or can you use `$this->anything()`?

File: tests/unit/Lookup/MultiplexingEntityLookupTest.php
Comment (JeroenDeDauw) [inline tests/unit/Lookup/MultiplexingEntityLookupTest.php]: Such PHP 5.3!

File: src/Lookup/UnknownForeignRepositoryException.php
Comment (filbertkm) [inline src/Lookup/UnknownForeignRepositoryException.php]: repoNname?

File: src/Lookup/MultiplexingEntityLookup.php
Comment (manicki) [reply src/Lookup/MultiplexingEntityLookup.php]: Renamed to `DispatchingEntityLookup` for now. Could come up with a nice way to include the repository name in the name (`RepositoryNameBasedDispatchingEntityLookup` is not the nicest name) but any suggestions are welcome!

**Comment by manicki** [comment]
I've done what @JeroenDeDauw and @jakobw suggested (except changing the array syntax, will do this separately for the whole repository)

File: tests/unit/Lookup/MultiplexingEntityLookupTest.php
Comment (manicki) [reply tests/unit/Lookup/MultiplexingEntityLookupTest.php]: True but this seems to be syntax used in this repository. I am always confused what Wikibase repo uses what array syntax. I'll made a separate PR for changing this

File: src/Lookup/UnknownForeignRepositoryException.php
Comment (manicki) [reply src/Lookup/UnknownForeignRepositoryException.php]: Forgot about this one. Thanks for spotting. Fixed now

**Comment by JeroenDeDauw** [comment]
Should this still be DNM?

**Comment by manicki** [comment]
maybe not but it still requires DM 6.2 release, so in this sense it is still kinda DNM. But removing the "label" now.

**Comment by JeroenDeDauw** [comment]
Ah right. You don't really need that if the tests are not passing :)

File: src/Lookup/DispatchingEntityLookup.php
Comment (jakobw) [inline src/Lookup/DispatchingEntityLookup.php]: Just like in the other PR this should probably also be `$entityId->getRepositoryName()` instead

Comment (manicki) [reply src/Lookup/DispatchingEntityLookup.php]: oops, right. Thanks!

**Comment by manicki** [comment]
Code adding the new exception have been separated to https://github.com/wmde/WikibaseDataModelServices/pull/154 so other PRs could also use it!

**Comment by manicki** [comment]
I've just realized that what I was trying to do here was wrong. Simple passing of things like `foo:Q123` to the lookup related to the repository locally known as `foo` does not seem right. That repo's database etc does not know a thing that some other repo calls it a `foo`.
What I believe should happen is that the dispatching lookup picks a right repo-specific lookup (that was done already in this patch), and then passes an entity id without a repo prefix to the right lookup.
I've added https://github.com/wmde/WikibaseDataModelServices/pull/150/commits/6669deb1433fa742cbc42980a19c2f504bf63d2d to do this.

Some "alternative" would be not to do this but rather adjust lookup code to not rely on the full id serialization but the form of it not including repo prefix. That seems logically incorrect for me, and also it would mean change some billion and a half places in Wikibase-related code to use `EntityId::getLocalPart` instead of `EntityId::getSerialization` so I decided not to go this path

**Comment by jakobw** [comment]
> Some "alternative" would be not to do this but rather adjust lookup code to not rely on the full id serialization but the form of it not including repo prefix. That seems logically incorrect for me, and also it would mean change some billion and a half places in Wikibase-related code to use EntityId::getLocalPart instead of EntityId::getSerialization so I decided not to go this path

Could you expand on why it would be logically incorrect to keep the repo prefix? To me it seems sensible to pass foreign EntityIds to services that operate on foreign repositories. It sounds like a dangerous thing to convert them to "local" early on. This would make it impossible for the Lookup-service to know whether the EntityId it received really belongs to its repo. Not sure if that's really a concern though.

**Comment by manicki** [comment]
> Could you expand on why it would be logically incorrect to keep the repo prefix? To me it seems sensible to pass foreign EntityIds to services that operate on foreign repositories. It sounds like a dangerous thing to convert them to "local" early on. 

For me `foo:Q123` reads like "an entity with id Q123 coming from the foreign repository I call foo". Following that, what I want to look up from the that repository (repo foo) is an entity identified as Q123. "foo" prefix only tells me I should be asking that repo, not my local repo.

Those repo prefixes only have a meaning inside the repo defining it. I believe it is not right to rely on them and refer to them outside of the scope of the "local" repo in any way.

> This would make it impossible for the Lookup-service to know whether the EntityId it received really belongs to its repo. Not sure if that's really a concern though.

I get this concern. But what if repo I call "foo" was using entities from some other repo, and by accident it would call the other repo also "foo". Then when my repo asks my "foo" for "foo:Q123" how the "my foo" repo knows is the question about its own Q123 or maybe the Q123 it got from the other foreign repo?

This whole process should be also taking the "prefix chaining" and "prefix mapping" into account. If repo I call "foo" sends me an entity "bar:Q321", and I have a mapping of "foo:bar" to "wd", I cannot ask repo foo for "wd:Q321" as it would not know if this is the request about its local Q321 or maybe the thing it knows as "bar:Q321" (to make the example even more horrible: let's assume repo foo used some entities from some other foreign repo and used "wd" for the prefix...). In such cases I think my repo should rather resolve "wd:Q321" first, and then ask repo "foo" for "bar:Q321".

**Comment by jakobw** [comment]
Ah, I think we were misunderstanding each other. Of course we should not ask "foo" what `foo:Q123` is. The prefix definitely has to be removed at some point. My question was whether it should be the job of the dispatching service or the service that the dispatcher picked. To my mind it makes more sense to make it the responsibility of the service that actually interacts with "foo" and not the dispatcher but there might be good reasons to do it the other way.

**Comment by manicki** [comment]
Oh, that's what you meant. Sorry I didn't get it.
You're making a good point actually. It indeed seems that there should be a decision that either all services should be able to handle the entity ids, no matter if they're foreign or not, or we assume that only _dispatching_ services deal with the foreign ids and always pass the "unprefixed" id to the right service.
At first glance I agree that it seems (to me) that the former would be nicer. On the other hand, that would most probably involve quite some change to many services. Not that should be seen as an argument against choosing that way.
I'd be happy to remove this change from this patch. It would be still worth looking deeper also into services we haven't touched yet. Some other opinions would be also appreciated.

It looks like would would also need a service that would handle the dropping of a prefixing and (especially) handle the prefix unmapping, and depending on the decision, use it in _all_ services or in the dispatching services.

**Comment by brightbyte** [comment]
Regarding the removal of the prefix: it's indeed not a trivial question. It boils down to whether the repo-specific service instances should "think like" the foreign repo, or like the local repo. Let's call this the "prefix resolution context".

Pragmatically, I propose the following:
- The dispatching services does not strip the prefix, it passes the EntityId to the underlying service as-is. This way, we can use the same API service instance for all repos we want to access via API, if the implementation looks up the API URL based on the prefix. This also allows us to stack multiple dispatching services (though I'm not sure we need this).
- service concrete implementations may be bound to a specific repo (they know a database name or API URL). Such implementations should check the prefix, and fail if expectations are not met.
- There could be a wrapper implementation of the service that does prefix stripping and/or un-mapping. If the service methods returns any EntityIds, prefix mapping should be applied, to maintain symmetry.
- A DB based implementation may only support local IDs, since it uses the serialized IDs as field values. To use that implementation with the DB of a foreign repo, it would need to be wrapped, to remove prefixed from input, and add prefixes to the output.
- when stripping prefixes, only the first part can be stripped. When processing "foo:bar:Q5", "foo" should be asked for "bar:Q5".

The mapping/unmapping wrapper gives us full flexibility. But it's quite a bit of boiler-plate code for each service interface. We should only do this if we have an implementation that is not "prefix-aware". The documentation of the interface should mention that implementations must clearly document whether they use the local context for prefixes, or a foreign one, or whether they support no prefixes.

File: src/Lookup/DispatchingEntityLookup.php
Comment (brightbyte) [inline src/Lookup/DispatchingEntityLookup.php]: I see no reason to require this here. The dispatcher shouldn't know or care that the empty prefix is somehow special.

To protect against bad configuration, the code that calls the constructor can do this check. That's bootsrap code that knows about configuration and the context in which the instance will be used. So it can know whether a handler for local IDs is needed.

File: src/Lookup/DispatchingEntityLookup.php
Comment (brightbyte) [inline src/Lookup/DispatchingEntityLookup.php]: Maybe we want a convenience function for this in some helper somewhere.

File: src/Lookup/DispatchingEntityLookup.php
Comment (brightbyte) [inline src/Lookup/DispatchingEntityLookup.php]: The dispatcher should not strip prefixes, see my post to the conversation thread. 
Since it's not obvious what the caller can expect, the documentation of this class should be explicit about whether prefixes are stripped or not.

File: src/Lookup/DispatchingEntityLookup.php
Comment (brightbyte) [inline src/Lookup/DispatchingEntityLookup.php]: ...based on the name of the repository an EntityId belongs to.

File: src/Lookup/DispatchingEntityLookup.php
Comment (brightbyte) [inline src/Lookup/DispatchingEntityLookup.php:60]: If we do prefix stripping above, the ID _inside_ the Entity would need to be changed to include the prefix now. It's important to maintain symmetry in this regard, so the caller only has to deal with the local "prefix resolution context".

**Review by brightbyte — CHANGES_REQUESTED** [review]
Dispatcher should not strip prefixes

File: src/Lookup/DispatchingEntityLookup.php
Comment (manicki) [reply src/Lookup/DispatchingEntityLookup.php]: done

File: src/Lookup/DispatchingEntityLookup.php
Comment (manicki) [reply src/Lookup/DispatchingEntityLookup.php]: Removed the requirement for the empty-string key

File: src/Lookup/DispatchingEntityLookup.php
Comment (manicki) [reply src/Lookup/DispatchingEntityLookup.php]: I've submitted something for this as https://github.com/wmde/WikibaseDataModelServices/pull/155.
To avoid creating yet more mess with all commits in this PR, I'd rather wait for that PR to at least be in shape ready to be merged and then do changes here (and in other places). 

Please do not "approve the change" (do not remove -1 in github's speak) before it is done so we don't forget about this.

File: src/Lookup/DispatchingEntityLookup.php
Comment (manicki) [reply src/Lookup/DispatchingEntityLookup.php]: Removed the code stripping prefixes

**Comment by brightbyte** [comment]
The code looks good now. As requested, I'm holding back on approving/merging until #155 is merged, and can be used here.

Also, please be aware that the "inner" EntityLookup must use the correct ID mapping when unserializing entities from another repository. This probably needs to be implemented in WikiPageEntityRedirectLookup / EntityContentDataCodec.

**Comment by addshore** [comment]
As with my 2p on https://github.com/wmde/WikibaseDataModelServices/pull/149 I have the same 2p here, that both classes and methods should have @since tags. 4 years down the line this makes everyone a bunch easier! and all for only the cost of a few lines!

**Comment by manicki** [comment]
Added since tags to public methods as well, and switched the new class to use validation function from DM 6.3.
As this PR has been getting horrendous I have squashed 20+ commits into one.
To verify I only changed two things mentioned above, one could run `git diff -r 5b8bb0a78c1656d24c4fd7bb1ae406ae0b265843 a8d7457a5c32fdd21a8a1f7c6cce5be69d130bfe` locally (watch out the rebase-related noise)

**Review by brightbyte — APPROVED** [review]

**Pull request merged**
